[feat] Gossip/SlowMo (#378)
facebookresearch/fairscale: docs/source/tutorials/slowmo_ddp.rst

Efficient Data Parallel Training with SlowMo Distributed Data Parallel
======================================================================

SlowMo Distributed Data Parallel reduces the communication between different
nodes while performing data parallel training. It is mainly useful for use on
clusters with low interconnect speeds between different nodes. When using
SlowMo, the models on the different nodes are no longer kept in sync after each
iteration, which leads to the optimization dynamics being affected. The end
result is close to the results of Distributed Data Parallel, but is not exactly
the same.

If you have code that is setup to use Distributed Data Parallel, using SlowMo Distributed Data Parallel
is simply replacing the DDP call with a call to

> ``fairscale.experimental.nn.data_parallel.SlowMoDistributedDataParallel``, adding a
> ``model.perform_slowmo(optimizer)`` call after ``optimizer.step()``, and moving the
> ``model.zero_grad(set_to_none=True)`` to be after ``optimizer.step()``, as follows.

Reviewer: ```suggestion
``fairscale.experimental.nn.data_parallel.SlowMoDistributedDataParallel``, and adding a
```suggestion
``model.perform_slowmo(optimizer)`` call after ``optimizer.step()`` -- preceded by
``model.zero_grad(set_to_none=True)`` in order to reduce peak memory usage.
`````````

The different points at which ``use_slowmo`` is used below help demonstrate these changes:

.. code-block:: python

    import torch
    from fairscale.experimental.nn.data_parallel import SlowMoDistributedDataParallel as SlowMoDDP

    def train(
        rank: int,
        world_size: int,
        epochs: int,
        use_slowmo: bool):

        # process group init
        dist_init(rank, world_size)

        # Problem statement
        model = MyAwesomeModel().to(rank)
        if use_slowmo:
            # Wrap the model into SlowMoDDP
            model = SlowMoDDP(model, slowmo_momentum=0.5, nprocs_per_node=8)
        else:
            model = DDP(model, device_ids=[rank])

        dataloader = MySuperFastDataloader()
        loss_ln = MyVeryRelevantLoss()
        optimizer = MyAmazingOptimizer()

>         # Any relevant training loop, with a line at the very end specific to SlowMoDDP. For example:

Reviewer: An unrelated suggestion, just to improve formatting in the docs as otherwise it spreads over two lines (at least on my laptop):
```suggestion
        # Any relevant training loop, with a line at the very end specific to SlowMoDDP, e.g.:
```

        model.train()
        for e in range(epochs):
            for (data, target) in dataloader:
                data, target = data.to(rank), target.to(rank)
                # Train
                if not use_slowmo:
                    model.zero_grad()
                outputs = model(data)
                loss = loss_fn(outputs, target)
                loss.backward()
                optimizer.step()
                if use_slowmo:
                    model.zero_grad(set_to_none=True)

>                     model.perform_slowmo(optimizer)  # SlowMoDDP specific

Reviewer: Also we can remove the comment here since the `if` block makes it clear it is SlowMo-specific
```suggestion
                    model.perform_slowmo(optimizer)
```

Efficient Data Parallel Training with SlowMo Distributed Data Parallel
======================================================================

SlowMo Distributed Data Parallel reduces the communication between different

> nodes while performing data parallel training. Let's suppose that your trainer

Reviewer 2: Does this look okay? Or, did you have other tradeoffs in mind?

> nodes while performing data parallel training. Let's suppose that your trainer

Reviewer: I believe it should at least mention that since the weights of the models on different nodes aren't kept in synch anymore after each iteration, the optimization dynamics are affected and this can lead to different results compared to DDP.

> nodes while performing data parallel training. Let's suppose that your trainer

Reviewer 2: Fixed this thanks

> nodes while performing data parallel training. Let's suppose that your trainer

Reviewer: Looks good, just `but Let's suppose that your trainer looks like` -> `Let's suppose that your trainer looks like`

> looks like -

Reviewer 2: Ahh so a large part of this tutorial is modified from other tutorials in fairscale. So I just went with what was previously there. It definitely makes sense to make these corrections though

`model.perform_slowmo(optimizer)` call after `optimizer.step()`, as follows.

.. code-block:: python

    import torch
    from fairscale.experimental.nn.data_parallel import SlowMoDistributedDataParallel as SlowMoDDP

    def train(
        rank: int,
        world_size: int,
        epochs: int):

        # process group init
        dist_init(rank, world_size)

        # Problem statement
        model = myAwesomeModel().to(rank)
        model = SlowMoDDP(model, nprocs_per_node=8)  # Wrap the model into SlowMoDDP
        dataloader = mySuperFastDataloader()
        loss_ln = myVeryRelevantLoss()
        optimizer = myAmazingOptimizer()

        # Any relevant training loop, with a line at the very end specific to SlowMoDDP. For example:
        model.train()
        for e in range(epochs):
            for (data, target) in dataloader:
                data, target = data.to(rank), target.to(rank)
                # Train
                model.zero_grad()
                outputs = model(data)
                loss = loss_fn(outputs, target)
                loss.backward()
                optimizer.step()
                model.perform_slowmo(optimizer)  # SlowMoDDP specific

> SlowMoDDP takes in slowmo_momentum as a parameter. This parameter may need to

Reviewer 2: 1. I have fixed this. Please let me know your thoughts on it
2 and 3. Thanks for these comments. I have made changes to fix these parts

> SlowMoDDP takes in slowmo_momentum as a parameter. This parameter may need to

Reviewer: Oh, also, you are using a single backtick for monospace text but my linter tells me you should use two in RST => does it actually give the proper formatting right now?

> SlowMoDDP takes in slowmo_momentum as a parameter. This parameter may need to

Reviewer 2: So I was getting a linter warning as well but it was still giving the proper formatting. Nevertheless, I have made a couple of changes to remove the linter warnings. However, there is still one linter warning coming, but I am thinking of ignoring that for now

> SlowMoDDP takes in slowmo_momentum as a parameter. This parameter may need to

Reviewer: Are you talking about `use_slowmo`? I see no reason not to use double backticks for it as well.

> SlowMoDDP takes in slowmo_momentum as a parameter. This parameter may need to

Reviewer 2: Ohh that's definitely right. Fixed it as well

        dist_init(rank, world_size)

        # Problem statement
        model = MyAwesomeModel().to(rank)
        model = DDP(model, device_ids=[rank])
        dataloader = MySuperFastDataloader()
        loss_ln = MyVeryRelevantLoss()
        optimizer = MyAmazingOptimizer()

        # Any relevant training loop, nothing specific to SlowMoDDP
        # For example:
        model.train()
        for e in range(epochs):
            for (data, targets) in dataloader:
                data, targets = data.to(rank), targets.to(rank)
                # Train
                model.zero_grad()
                outputs = model(data)
                loss = loss_fn(outputs, targets)
                loss.backward()
                optimizer.step()

Then using SlowMo Distributed Data Parallel is simply replacing the DDP call with a call to
`fairscale.experimental.nn.data_parallel.SlowMoDistributedDataParallel` and adding a
`model.perform_slowmo(optimizer)` call after `optimizer.step()`, as follows. The different

> points at which `use_slowmo` is used below helps demonstrate these changes.

Reviewer: ```suggestion
points at which `use_slowmo` is used below help demonstrate these changes.
```

            for (data, target) in dataloader:
                data, target = data.to(rank), target.to(rank)
                # Train
                if not use_slowmo:
                    model.zero_grad()
                outputs = model(data)
                loss = loss_fn(outputs, target)
                loss.backward()
                optimizer.step()
                if use_slowmo:

>                     model.zero_grad()

Reviewer: I assume it is important to have it here for SlowMo, but I don't remember why: it would be good to explain it in the docstring of `perform_slowmo()` and refer to this doc here.

Reviewer: Thanks for the update, a couple of follow-up comments on this point:

1. Minor: would it be possible for the doc link to directly point to the `perform_slowmo` part of the page? (no big deal if not possible)

2. How does this save memory? According to the doc (https://pytorch.org/docs/stable/generated/torch.nn.Module.html) it won't flush the tensors unless `set_to_none` is set to True

Reviewer 2: 1. Have fixed this. The link is a little ugly but it has very little chance of breaking in the future, so it might be good to go ahead with

2. Ahh nice catch, I've fixed that. In the fairseq repo, setting to None was the default behavior of zero_grad so I got confused about that

                    model.perform_slowmo(optimizer)  # SlowMoDDP specific

In the example above, when using SlowMoDDP, we are reducing the total communication between

> nodes by 3 times as the default ``localsgd_frequency`` is set to 3 by default.

Reviewer: `default` is repeated twice here, one should be removed
